### PR TITLE
Use AdoptOpenJDK JDK 8 for testing [ci: last-only]

### PR DESCRIPTION
**Pull request by eed3si9n**
Ref https://github.com/scala/scala-dev/issues/587

### CODE REVIEW

**Comment by retronym** [comment]
Rationale?

**Comment by eed3si9n** [comment]
AdoptOpenJDK is the most consistently available distribution of OpenJDK code base, so that's what Lightbend open source projects are adopting by policy.

**Comment by adriaanm** [comment]
how long does the jdk install take? can we cache some of that?

**Comment by eed3si9n** [comment]
It's cached:

```
 - $HOME/.sdkman
```
It should be like zero seconds from the second run onwards.

File: .travis.yml [hunk 1]
```diff
@@ -63,6 +77,7 @@ jobs:
 
 env:
   global:
+    - TRAVIS_JDK=8.0.202.hs-adpt
```
Comment (retronym) [inline .travis.yml]: I don't really like baking this exact version number into `travis.yml`. Typically, we just ask for 8/9/12 and let external config in the CI platform pick the latest minor release for that.

Comment (retronym) [reply .travis.yml]: Travis's native support for JDK installation comes from:

https://github.com/sormuras/bach/blob/master/install-jdk.sh

Should we contribute AdoptOpenJDK support to that script?

Comment (eed3si9n) [reply .travis.yml]: We can do

```
sdk list java | grep -o "8\..*hs-adpt" | tail -1
```

to let sdkman decide the patch version.

Comment (eed3si9n) [reply .travis.yml]: If the bach maintainer is willing to maintain AdoptOpenJDK links, that would be great.
See also
- https://github.com/travis-ci/travis-cookbooks/issues/976
- https://github.com/travis-ci/travis-ci/issues/9368
- https://github.com/travis-ci/travis-build/pull/1347

Comment (retronym) [reply .travis.yml]: Looks like `install_jdk.sh` already supports Adopt OpenJDK: 

Example: https://github.com/sormuras/sormuras.github.io/blob/master/.travis.yml#L27

The only catch is that it doesn't have an alias for it that can be specified in the `jdk: openjdk8` directive in `.travis.yml`.

This URL resolves to the latest release of AdoptOpenJDK 8: https://api.adoptopenjdk.net/v2/binary/releases/openjdk8?openjdk_impl=hotspot&os=linux&arch=x64&release=latest&heap_size=normal&type=jdk

Comment (eed3si9n) [reply .travis.yml]: Added patch detection here - b45f9aa61d553b94e10fa297017c22455d3f9fe0

File: .travis.yml [hunk 2]
```diff
@@ -2,13 +2,26 @@
 sudo: required
```
Comment (SethTisue) [inline .travis.yml]: omit, and maybe add `dist:` line, like you did in the scala-xml PR, so Travis doesn't just pick one?

**Review by SethTisue — CHANGES_REQUESTED** [review]
should this target 2.12.x rather than 2.13.x?

otherwise LGTM

**Comment by SethTisue** [comment]
re: "rationale?", umbrella ticket is https://github.com/scala/scala-dev/issues/587

File: .travis.yml [hunk 3]
```diff
@@ -1,14 +1,27 @@
...
+before_install:
+  # adding $HOME/.sdkman to cache would create an empty directory, which interferes with the initial installation
+  - "[[ -d /home/travis/.sdkman/ ]] && [[ -d /home/travis/.sdkman/bin/ ]] || rm -rf /home/travis/.sdkman/"
+  - curl -sL https://get.sdkman.io | bash
+  - echo sdkman_auto_answer=true > /home/travis/.sdkman/etc/config
+  - source "/home/travis/.sdkman/bin/sdkman-init.sh"
+
+install:
+  - sdk install java $(sdk list java | grep -o "$ADOPTOPENJDK\.[0-9\.]*hs-adpt" | tail -1)
+  - unset _JAVA_OPTIONS
```
Comment (dwijnand) [inline .travis.yml]: This is no longer necessary

File: .travis.yml [hunk 4]
```diff
@@ -1,14 +1,27 @@
...
 cache:
   directories:
     - $HOME/.ivy2/cache
     - $HOME/.sbt
     - $HOME/.rvm/
+    - $HOME/.sdkman
+
+before_install:
+  # adding $HOME/.sdkman to cache would create an empty directory, which interferes with the initial installation
+  - "[[ -d /home/travis/.sdkman/ ]] && [[ -d /home/travis/.sdkman/bin/ ]] || rm -rf /home/travis/.sdkman/"
```
Comment (dwijnand) [inline .travis.yml]: Should we stick to `$HOME/.sdkman` seeing as those that's what's used in `cache.directory`?

(Also you can drop the `[[ -d /home/travis/.sdkman/ ]]` check and just check for `[[ -d "$HOME/.sdkman/bin/" ]]`.)

File: .travis.yml [hunk 5]
```diff
@@ -1,14 +1,27 @@
...
+  - curl -sL https://get.sdkman.io | bash
+  - echo sdkman_auto_answer=true > /home/travis/.sdkman/etc/config
+  - source "/home/travis/.sdkman/bin/sdkman-init.sh"
+
+install:
+  - sdk install java $(sdk list java | grep -o "$ADOPTOPENJDK\.[0-9\.]*hs-adpt" | tail -1)
```
Comment (dwijnand) [inline .travis.yml]: I'm not so sure about this implementation as it seems `sdk list` wasn't built to be scripted around.

Particularly the `tail -1` it seems to me to be there to deal with the `sdk install java 11.0.3.hs-adpt` example at the bottom:

```
08:17:43 $ sdk list java
================================================================================
Available Java Versions
================================================================================
 Vendor        | Use | Version      | Dist    | Status     | Identifier
--------------------------------------------------------------------------------
 AdoptOpenJDK  |     | 12.0.1.j9    | adpt    |            | 12.0.1.j9-adpt
               |     | 12.0.1.hs    | adpt    |            | 12.0.1.hs-adpt
               |     | 11.0.3.j9    | adpt    |            | 11.0.3.j9-adpt
               |     | 11.0.3.hs    | adpt    |            | 11.0.3.hs-adpt
               |     | 8.0.212.j9   | adpt    |            | 8.0.212.j9-adpt
               |     | 8.0.212.hs   | adpt    |            | 8.0.212.hs-adpt
...
================================================================================
Use the Identifier for installation:

    $ sdk install java 11.0.3.hs-adpt
================================================================================
```

It looks like SDKMAN! only tracks the latest patch version of major versions, so if anything `head -1` is probably better, to get the latest in the list, not the version used as an example.

```suggestion
  - sdk install java $(sdk list java | grep -o "$ADOPTOPENJDK\.[0-9\.]*hs-adpt" | head -1)
```

Alternatively should we use SDKMAN!'s Candidates Service (https://github.com/sdkman/sdkman-candidates) API?   Assuming that's a reliably stable API we could use:

```suggestion
  - sdk install java $(curl -Ls "$SDKMAN_CANDIDATES_API/candidates/java/$SDKMAN_PLATFORM/versions/all" | tr ',' '\n' | grep "$ADOPTOPENJDK\.[0-9\.]*\.hs-adpt")
```

which behaves like this:

```bash
08:29:29 $ for ADOPTOPENJDK in 8 11; do
> curl -Ls "$SDKMAN_CANDIDATES_API/candidates/java/$SDKMAN_PLATFORM/versions/all" | tr ',' '\n' | grep "$ADOPTOPENJDK\.[0-9\.]*\.hs-adpt"
> done
8.0.212.hs-adpt
11.0.3.hs-adpt
```

File: .travis.yml [hunk 6]
```diff
@@ -1,14 +1,27 @@
...
+  - unset _JAVA_OPTIONS
+  - java -Xmx32m -version
+  - javac -version
```
Comment (dwijnand) [inline .travis.yml]: While we're at it 🙂

```suggestion
  - javac -J-Xmx32m -version
```

**Review by dwijnand — CHANGES_REQUESTED** [review]
Looks great.  I agree and like how this puts us back in a situation where we use the latest patch version of the JDK (assuming SDKMAN!'s metadata stays up to date).

I have a couple of mid-sized questions/requests.

File: .travis.yml (hunk 5 above)
Comment (dwijnand) [reply .travis.yml]: I reached out to the SDKMAN! project, which prefers that initial points of contact happen on their Slack, so the link to my thread on their Slack is https://sdkman.slack.com/archives/CJK9DMV6V/p1562658566004100.

Comment (dwijnand) [inline .travis.yml]: Alternatively, we could stick to using `install-jdk.sh` which (apparently, but we should verify) was just updated by Travis CI (https://github.com/sormuras/bach/issues/62), by adding the following as an `install` step:

```yaml
- source install-jdk.sh --url "https://api.adoptopenjdk.net/v2/binary/releases/openjdk$ADOPTOPENJDK?openjdk_impl=hotspot&os=linux&arch=x64&release=latest&heap_size=normal&type=jdk"
```

This is where those URLs redirect to:

```
09:11:36 $ for ADOPTOPENJDK in 8 11; do
> curl -I "https://api.adoptopenjdk.net/v2/binary/releases/openjdk$ADOPTOPENJDK?openjdk_impl=hotspot&os=linux&arch=x64&release=latest&heap_size=normal&type=jdk"
> done
HTTP/2 302
date: Tue, 09 Jul 2019 08:19:37 GMT
content-type: text/plain; charset=utf-8
content-length: 150
set-cookie: __cfduid=dbe707790dae1c34dc2e23ec1cdf42a581562660376; expires=Wed, 08-Jul-20 08:19:36 GMT; path=/; domain=.adoptopenjdk.net; HttpOnly
x-powered-by: Express
access-control-allow-origin: *
location: https://github.com/AdoptOpenJDK/openjdk8-binaries/releases/download/jdk8u212-b04/OpenJDK8U-jdk_x64_linux_hotspot_8u212b04.tar.gz
vary: Accept
set-cookie: b7b892882bae631693e1ea44963ef628=afdd0ec2d9aff4e059f50dfa0530835e; path=/; HttpOnly; Secure
expect-ct: max-age=604800, report-uri="https://report-uri.cloudflare.com/cdn-cgi/beacon/expect-ct"
server: cloudflare
cf-ray: 4f38dfbb3ff9dbdb-LHR

HTTP/2 302
date: Tue, 09 Jul 2019 08:19:37 GMT
content-type: text/plain; charset=utf-8
content-length: 154
set-cookie: __cfduid=dab5451e461b1bce70eeaab3f5054a6981562660377; expires=Wed, 08-Jul-20 08:19:37 GMT; path=/; domain=.adoptopenjdk.net; HttpOnly
x-powered-by: Express
access-control-allow-origin: *
location: https://github.com/AdoptOpenJDK/openjdk11-binaries/releases/download/jdk-11.0.3%2B7/OpenJDK11U-jdk_x64_linux_hotspot_11.0.3_7.tar.gz
vary: Accept
set-cookie: b7b892882bae631693e1ea44963ef628=afdd0ec2d9aff4e059f50dfa0530835e; path=/; HttpOnly; Secure
expect-ct: max-age=604800, report-uri="https://report-uri.cloudflare.com/cdn-cgi/beacon/expect-ct"
server: cloudflare
cf-ray: 4f38dfbf3df9ce2f-LHR
```

Comment (dwijnand) [reply .travis.yml]:
```yaml
before_install: wget https://github.com/sormuras/bach/raw/master/install-jdk.sh
```

Comment (lrytz) [reply .travis.yml]: +1

If that doesn't work, jabba supports installing from an url.

```
jabba install 1.8-adopt-latest=tgz+'https://api.adoptopenjdk.net/v2/binary/releases/openjdk8?openjdk_impl=hotspot&os=linux&arch=x64&release=latest&heap_size=normal&type=jdk'
```

works on my machine.

Comment (lrytz) [reply .travis.yml]: Since we cache the JDK, we should also check that the cached version is still the latest.

File: .travis.yml (hunk 4 above)
Comment (eed3si9n) [reply .travis.yml]: I sort of remember `$HOME` not always working in some parts of the travis yaml. Maybe due to the initialization ordering these scripts are executed? I don't have specific memory on this being one of those situations, or if I was just being paranoid.

Comment (dwijnand) [reply .travis.yml]: Happy to change these to `$HOME`?

Comment (eed3si9n) [reply .travis.yml]: Yea. I am happy to give it a shot again.

Comment (eed3si9n) [reply .travis.yml]: 5f04df40618f9213e36be38d687a7ba550007afd

File: .travis.yml (hunk 5 above)
Comment (eed3si9n) [reply .travis.yml]: here's using head - https://github.com/scala/scala/pull/7931/commits/5f04df40618f9213e36be38d687a7ba550007afd

**Comment by eed3si9n** [comment]
This is re-targeted to 2.12.x, and squashed.

File: .travis.yml [hunk 7]
```diff
@@ -1,14 +1,18 @@
-# GCE VMs have better performance (will be upgrading to premium VMs soon)
-sudo: required
-
+dist: xenial
+group: stable
```
Comment (lrytz) [inline .travis.yml:2]: so did we decide to keep those two entries in the end?

Comment (eed3si9n) [reply .travis.yml:2]: I did attempt to removed it, but for some reason Travis CI picked up trusty - https://travis-ci.org/scala/scala/jobs/556437704#L7-L9

Comment (lrytz) [reply .travis.yml:2]: OK. Thanks!

**Pull request merged**